Feat/form wizard

9helix opened the pull request:
## Description

Added functional component Wizard for creating dynamic form-filling process.
## Change summary

- Added wizard component under `organisms`
- Added example usage in `src/app/page.tsx`

## Checklist

- [x] I have performed a self-review of my code
- [x] I tested locally
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New package or package update introduced

aarbanas commented on src/app/page.tsx:
```diff
@@ -1,11 +1,59 @@
+  return (
+    <>
+      <Users />
+      <Wizard steps={["Sign Up", "Details", "Billing"]} inputs={inputs} />
```
This approach can cause multiple issues. 
Developer can define `steps` array with one length and `inputs` array with different length which can break the `Wizard` component. 
I can see two solutions for this.
1. This one can be easier to implement is to define inputs as `Record<string, WizardInputValue`. Like that input will have both the keys and values which need to be passed to the `Wizard` component. (Example at the end)
2. Create validation for the current implementation that typescript is complaining at compile time if `steps.length != inputs.length`


Example for solution one
```typescript
type WizardInputValue = {
  type: 'text' | 'number' | 'data' | 'select' | 'checkbox',
  label: string,
  placeholder: string
}

interface WizardProps {
  inputs: Record<string, WizardInputValue>
}
```

9helix replied:
I prefer the first approach as it prevents the initial problem rather than fixing it. But I don't think `Record` is a suitable type as we need our steps to be ordered. Here I used an array instead:

```ts
interface Fields {
  type: "text" | "password" | "email" | "number" | "tel" | "url" | "search";
  label: string;
  placeholder: string;
}

interface FormStep {
  name: string;
  fields: Fields[];
}

interface FormProps {
  form: FormStep[];
}
```
What do you think?

aarbanas replied:
I like it 🚀

IMilja replied:
I like it as well, but one thing that I would like to add as a suggestion is that we add a `name` property to the `Fields` type as well since the `name` is the main part that is being used to know which field is which when submitting a form.

IMilja replied:
Also, one thing I remembered, but this approach could cause us some issues with the ability to add values to a `select` element if we're required to add e.g. "Select your country" to the form.

aarbanas commented on src/components/organisms/wizard/Wizard.tsx:
```diff
@@ -0,0 +1,109 @@
+  );
+};
+
+export default Wizard;
```
Maybe to think of some more explicit component name than `Wizard`. When you see somewhere in the code component `Wizard` it is really hard to tell what it is doing and what it is meant to be used for. Something like `CreationWizard` is saying more, but TBH I do not like it as well. @IMilja do you have any better ideas?

9helix replied:
`MultiStepForm`?

aarbanas replied:
`MultistepFormWizard`? @IMilja your vote 😂

IMilja replied:
My vote is going towards `MultistepForm` cause it makes to most sense to me. The word `Wizard` and the scope of what component does not really "match".

aarbanas reviewed (commented):
Great job! Just leaving some comments with my concerns

IMilja commented on src/components/organisms/wizard/Wizard.tsx:
```diff
@@ -0,0 +1,109 @@
+    e: React.ChangeEvent<HTMLInputElement>,
+    index: number,
+  ) => {
+    setFormData({
+      ...formData,
+      [`${currentStep}-${index}`]: e.target.value,
+    });
```
```suggestion
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
    });
```
Using [`${currentStep}-${index}\`] will make it super hard for us to send it to the BE later. Either way, we will use react-hook-form to submit the forms but it's something to keep in mind :).

IMilja commented on src/components/organisms/wizard/Wizard.tsx:
```diff
@@ -0,0 +1,109 @@
+              value={formData[`${currentStep}-${index}`] ?? ""}
+              onChange={(e) => handleChange(e, index)}
+              className="w-96"
+              label={input.label}
```
As mentioned in one of my previous comments regarding the `Forms` type. Could we add `name` to the field render?

9helix commented:
Now I've added an alternative `MultiStepForm` component which accepts subforms and then nests them inside of a `FormComponent`.

aarbanas commented on src/components/organisms/multiStepForm/MultiStepFormNEW.tsx:
```diff
@@ -0,0 +1,107 @@
+    //nextStep();
+    //console.log(typeof data);
+    if (currentStep == numSteps - 1 && isLastStep) alert(JSON.stringify(data));
+    // Add submission logic here
```
This logic must be provided from parent component. Because handle submit usually is calling some API. I can see two possible solutions.
1. Add another prop to `MultiStepForm`. Besides `form`, add `formSubmission` void function which is going to handle the submission process.
2. Change the `FormStep`.
   1. Rename it to `FormAction` 
   2. It should contain array of `FormSteps` and `onSubmit` function
 
```typescript
export type FormAction {
  formSteps: FormStep[];
  onSubmit: void();
};
 
export type FormStep {
  name: string;
  form: ReactElement;
}
```

IMilja commented on src/components/organisms/forms/Billing.tsx:
I agree with @aarbanas on this one. I would even push it further and say that you can add the Billing, Signup and Details forms into `app/users/_components` since all forms are user-related feel free to bundle them into the same folder.

@aarbanas Agree/Disagree? :eye: :eye:  :smile_cat:

9helix commented:
I've added that an `onSubmit` function can be provided as a prop to the `MultiStepForm`. But I had to use it from a Client component so I moved part of the logic from `src/app/page.tsx` into a separate file. I also moved form components to the `src/app/users/_components` folder.
I also noticed that `Inputs` type is in `MultiStepForm` component although it shouldn't be as field values depend on the used form components. I solved it by wrapping the component inside of a generic function.
So instead of the Inputs being in `MSF` component:
```ts
type Inputs = {
  username: string;
  password: string;
  password2: string;
  email: string;
  phone: string;
  creditCard: string;
  billingAddr: string;
};
``` 
 I did:
```ts
  interface Inputs extends UserCredentials, ContactInfo, PaymentInfo {}
  return generateForm<Inputs>(test, onSubmit);
  ```
  Inside of the `generateForm` function there is our `MSF` component.
  What do you think of this solution?

aarbanas reviewed (approved):
LGTM
